sys/exec.c

```diff
@@ -1,21 +1,162 @@
 #include <common.h>
 #include <stdc.h>
+#include <elf/mips_elf.h>
+#include <vm_map.h>
+#include <vm_pager.h>
+#include <thread.h>
 
 extern uint8_t _binary_prog_uelf_start[];
 extern uint8_t _binary_prog_uelf_size[];
 extern uint8_t _binary_prog_uelf_end[];
 
-void exec(){
-    size_t prog_size = (size_t)(void*)_binary_prog_uelf_size;
-    kprintf("[exec] User ELF size: %ld\n", prog_size);
+int exec(){
+    uint8_t* elf_image = _binary_prog_uelf_start;
+    size_t elf_size = (size_t)(void*)_binary_prog_uelf_size;
+
+    kprintf("[exec] User ELF size: %ld\n", elf_size);
+
+    if(elf_size < sizeof(Elf32_Ehdr)/sizeof(uint8_t)){
+        kprintf("[exec] Exec failed: ELF file is too small to contain a valid header\n");
+        return -1;
+    }
 
     const uint8_t header_size = 64;
     const uint8_t stride = 8;
     kprintf("[exec] User ELF header dump:\n");
     for(int8_t i = 0; i < header_size; i++){
         if(i % stride == 0) kprintf("[exec]  ");
-        uint8_t byte = _binary_prog_uelf_start[i];
+        uint8_t byte = elf_image[i];
         kprintf(" %02x", byte);
         if(i % stride == 7) kprintf("\n");
     }
+
+    const Elf32_Ehdr* eh = (Elf32_Ehdr*)elf_image;
+
+    // Start by determining the validity of the elf file.
+
+    // First, check for the magic header.
+    if(eh->e_ident[EI_MAG0] != ELFMAG0 ||
+       eh->e_ident[EI_MAG1] != ELFMAG1 ||
+       eh->e_ident[EI_MAG2] != ELFMAG2 ||
+       eh->e_ident[EI_MAG3] != ELFMAG3){
+        kprintf("[exec] Exec failed: Incorrect ELF magic number\n");
+        return -1;
+    }
+    // Check ELF class
+    if(eh->e_ident[EI_CLASS] != ELFCLASS32){
+        kprintf("[exec] Exec failed: Unsupported ELF class (!= ELF32)\n");
+        return -1;
+    }
+    // Check data format endianess
+    if(eh->e_ident[EI_DATA] != ELFDATA2LSB){
+        kprintf("[exec] Exec failed: ELF file is not low-endian\n");
+        return -1;
+    }
+    // Ignore version and os abi field
+    // Check file type
+    if(eh->e_type != ET_EXEC){
+        kprintf("[exec] Exec failed: ELF is not an executable\n");
+        return -1;
+    }
+    // Check machine architecture field
+    if(eh->e_machine != EM_MIPS){
+        kprintf("[exec] Exec failed: ELF target architecture is not MIPS\n");
+        return -1;
+    }
+
+    // Take note of the entry point
+    kprintf("[exec] Entry point will be at 0x%08x.\n",
+            (unsigned int)eh->e_entry);
+
+    // Ensure prog header size compliance
+    if(eh->e_phentsize != sizeof(Elf32_Phdr)/sizeof(uint8_t)){
```

Comment:
This is stricter than necessary. ELF format is designed so that it can be extended while maintaining backward compatibility. For example, a new field to segment entries can be easily added and old implementations will still be able to load the file, just ignoring the new fields.

`sizeof(uint8_t)` seems redundant, it is going to be 1.